Release/3.0.0

paulboocock opened the pull request:
Oh, hello there 👋🏻 This is the intial PR for the Snowplow JavaScript Tracker v3 release. 

## Webinar - March 18th
We're hosting a product office hours and meetup on Thursday where I'll be running through the new tracker, the changes and a quick demo of using it in a react app. If that isn't exciting enough, during the meetup we'll have a special guest who is a shark scientist! Actual sharks! You can register here: https://snowplowanalytics.com/events/product-office-hours/

## Highlights
- Available on [NPM](https://www.npmjs.com/package/@snowplow/browser-tracker) as `@snowplow/browser-tracker`. Install with `npm install @snowplow/browser-tracker@next` for the latest alpha/beta release.
- Now with configurable (and extendable) plugin architecture.
- Plugins also published on [NPM](https://www.npmjs.com/search?q=%40snowplow) (e.g. `@snowplow/browser-plugin-ad-tracking` and `@snowplow/browser-plugin-consent`).
- New track API, pass objects rather than parameter lists (e.g. `window.snowplow('trackPageView', { title: 'My Website', contexts: [ ... ] });`)
- [Published as traditional](https://github.com/snowplow/snowplow-javascript-tracker/releases) `sp.js` (66.7 KB - 21.2 KB gzipped) and new `sp.lite.js` (35.5 KB - 12.3 KB gzipped). `sp.lite.js` has Page View, Self Describing and Structured Events as well as Activity Tracking and Anonymous Tracking.
-`sp.js` and `sp.lite.js` are available on [jsDelivr](https://cdn.jsdelivr.net/npm/@snowplow/javascript-tracker@3.0.0-alpha.1/dist/) and [unpkg](https://unpkg.com/browse/@snowplow/javascript-tracker@3.0.0-alpha.1/dist/).
- Build your own `sp.js` by following the setup instructions and editing `/trackers/javascript-tracker/tracker.config.js`.
- This repo is now a fancy `@microsoft/rush` monorepo.
- No more lodash!
- Completely rewritten in TypeScript, shipped with full type support.
- Still supports IE9 and Safari 8 😬 
- Removed `f_` feature fields, Optimizely Classic and Parrable tracking from default sp.js (still available as plugins).

Take it for a spin and let us know what you think in the coments 🎉 

## Quick start
If you want to test the new browser-tracker along with some trackers, heres a couple of lines of code that should get you started

```typescript
import { newTracker, trackPageView } from "@snowplow/browser-tracker";
import { AdTrackingPlugin, trackAdClick } from "@snowplow/browser-plugin-ad-tracking";
import { PerformanceTimingPlugin } from "@snowplow/browser-plugin-performance-timing";

newTracker('sp1', '{{collector_url}}', { 
  appId: 'my-app-d', 
  plugins: [ PerformanceTimingPlugin(), AdTrackingPlugin() ],
  eventMethod: 'post'
})
trackPageView();

function trackAnEvent() {
  trackAdClick({targetUrl: 'https://www.snowplowanalytics.com', advertiserId: '1234-5678'});
}
```

If you're looking to test out the classic `sp.js` then you'll find them in the usual Github Releases section, tagged as Pre-releases. There is a big breaking change between this version and v2, and that is all functions now accept an object rather than a list of parameters. I've put a couple of examples below:

```javascript
      window.snowplow('trackPageView');

      window.snowplow('trackPageView', {
        title: 'My Title',
        context: [
          {
            schema: 'iglu:org.schema/WebPage/jsonschema/1-0-0',
            data: {
              keywords: ['tester'],
            },
          },
        ],
      });

      window.snowplow('trackSelfDescribingEvent', {
        event: {
          schema: 'iglu:com.snowplowanalytics.snowplow/ad_impression/jsonschema/1-0-0',
          data: {
            bannerId: 'ASO01043',
          },
        },
        context: [],
        timestamp: { type: 'ttm', value: 1477401868 },
      });

      window.snowplow('enableActivityTracking', { minimumVisitLength: 15, heartbeatDelay: 15 });
```

## TODO
- [x] Update/tidy code documentation and comments
- [x] Create rush `change` files
- [x] Release 3.0.0-alpha.1 to test rush publishing
- [x] Start publishing on NPM
- [x] Fix Tracker Version being 2.17.3 for Alpha and Beta releases
- [x] Release 3.0.0-alpha.2
- [x] Allow plugins to be loaded dynamically when using `sp.js`
- [x] Investigate #894, #884 and #662
- [x] Bump any dependencies
- [x] Complete #913
- [x] Action any alpha feedback
- [x] Release 3.0.0-beta.1
- [x] Action any beta.1 feedback
- [x] Release 3.0.0-beta.2
- [x] Incorporate node-tracker into monorepo
- [x] Release 3.0.0-beta.3
- [x] Testing
- [x] Publish API documentation with TSDoc
- [x] Write documentation
- [x] Release 3.0.0

mwdiaz commented:
Awesome! I'd love to try this out, but it appears it actually hasn't been published to npm yet?

```
[0] % npm info @snowplow/browser-tracker
npm ERR! code E404
npm ERR! 404 '@snowplow/browser-tracker' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it
npm ERR! 404 (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
npm ERR! 404
npm ERR! 404  '@snowplow/browser-tracker@latest' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
```

paulboocock commented:
Not yet, it's item 3 on the todo list. I expect it'll be published on Monday.

paulboocock commented:
It's now available! I've added a mini guide to the top comment, which I strongly advise reading before giving this a spin.

https://www.npmjs.com/package/@snowplow/browser-tracker

Known issue: `3.0.0-alpha.1` reports the tracker version as `js-2.17.3`.

paulboocock commented:
`3.0.0-alpha.3` is now available. This includes some improvements to event flushing and removes `pageUnloadTimer`. Also ensures all API interfaces are exported as types so you can use them in your own code.

The next release will be `3.0.0-beta.1` which I anticipate will drop on Monday. 

I'm also hosting a product office hours and meetup on Thursday where i'll be running through the new tracker, the changes and a quick demo of using it in a react app. If that isn't exciting enough, during the meetup we'll have a special guest who is a shark scientist! Actual sharks! You can register here: https://snowplowanalytics.com/events/product-office-hours/

paulboocock commented:
beta.1 is now available: https://discourse.snowplowanalytics.com/t/snowplow-javascript-tracker-v3-beta-release-webinar/4908

paulboocock commented:
3.0.0-beta.2 is now available. The changes from 3.0.0-beta.1:
- The tracker should now be importable in Next.js when using SSR without throwing `window is not defined`.
- The Parrable Plugin has been deprecated and removed following feedback in #925 
- Fixed some types on ecommerce cart events

paulboocock commented:
3.0.0-beta.3 is now available. The changes from 3.0.0-beta.2:

- We are now also publishing the Node.js Tracker from this monorepo repository as `@snowplow/node-tracker\
https://www.npmjs.com/package/@snowplow/node-tracker (Documentation updates coming soon)
- Now exporting the `BrowserTracker` and `TrackerConfiguration` interfaces from `@snowplow/browser-tracker`

jamessnowplow commented:
Hi there,

I'm just wondering about consistency around the `addPlugin` methods between node-tracker and browser-tracker.

The node version being a mirror of `TrackerCore.addPlugin(plugin: CorePlugin)` and the browser one having an extra wrapper around the passed plugin `addPlugin(configuration: { plugin: BrowserPlugin }, trackers?)`.

I dont think its very important and typescript will let you know whats going on, just worth a mention.

colmsnowplow commented on libraries/browser-tracker-core/src/tracker/index.ts:
```diff
@@ -0,0 +1,1247 @@
+        // Update currentVisitTs
+        idCookieComponents[4] = Math.round(new Date().getTime() / 1000);
+        idCookieComponents.shift();
+        setDomainUserIdCookie.apply(null, idCookieComponents as any); // TODO: Remove any
```
Is this comment left in intentionally?

paulboocock replied:
I did this because adding types to all the cookie logic was rather, uhm, tricky. I might have another go at it now you've reminded me 😉

colmsnowplow replied:
Cool cool just double checking it's not an oversight. :)

jamessnowplow reviewed (approved):
Loving how this has shaped up.

The plugin arrangement is a game changer and having a modern build setup around it sets things up wonderfully.

colmsnowplow reviewed (approved):
I totally agree with James, this is fantastic! Love how things have shaped up.

Have tested the traditional tag-style implementation, all works as expected.

One thing to point out is that the `trackStructEvent` method won't error if left the old way, but will send an event with all null fields. I don't think this is really a problem, just an FYI - all the others I tested will explicitly error if left the old way (I think by chance).

I love how easy it is to add plugins and build your own sp.js asset!
